src/os/bluestore/BlueFS.cc

```diff
@@ -2106,6 +2133,141 @@ int BlueFS::device_migrate_to_new(
+    // use normal v1 write path by marking node type to legacy
+    writer->file->fnode.type = bluefs_node_type::REGULAR;
+
+    FileReader* previous_reader;
+    r = open_for_read(wal_dir, name, &previous_reader);
+    auto read_writer_path = make_scope_guard([&] {
+      delete previous_reader;
+    });
+    ceph_assert(r == 0);
+    
+    
+    // move contents from previous to new file
+    uint64_t previous_size = previous_reader->file->fnode.wal_size;
+    uint64_t left_to_read = previous_size;
+    bufferlist read_buffer;
+    uint64_t read_offset = 0;
+    while(left_to_read > 0) {
+      // read from prevoius file
+      read_buffer.clear();
+      uint64_t to_read = std::min(buffer_size, left_to_read);
+      int r = _read_wal(previous_reader, read_offset, to_read, &read_buffer, nullptr);
+      ceph_assert(r >= 0);
+      if (uint64_t(r) < to_read) {
+        dout(5) << fmt::format("{} read less than expect while migrating file {} to wal v1", __func__, name) << dendl;
+        failure_while_migrating = true;
+        break;
+      }
+      
+      // add to new file
+      append_try_flush(writer, read_buffer.c_str(), read_buffer.length());
+      
+      // update state
+      read_offset += to_read;
+      left_to_read -= to_read;
+    }
+    if (failure_while_migrating) {
+      break;
+    }
+    // todo things left to read, do cleanup
+    
+    int fsync_ret = fsync(writer);
+    if (fsync_ret != 0) {
+      derr << fmt::format("{} fsync error {} while moving data for file {} ", __func__, fsync_ret, name) << dendl;
+      failure_while_migrating = true;
+      break;
+    }
+    
+    file_migrated_count++;
+  }
+  
+  if (failure_while_migrating) {
+    derr << fmt::format("{} couldn't properly move data from v2 to v1, unlinking temp files...", __func__) << dendl;
+    size_t file_migrated_count = 0;
+    for (size_t i = 0; i < files_to_migrate.size(); i++) {
+      string new_file_temp_name = fmt::format("{}{}", tmp_name_prefix, file_migrated_count);
+      derr << fmt::format("{} unlinking {}", __func__, new_file_temp_name) << dendl;
+      int unlink_ret = unlink(wal_dir, new_file_temp_name);
+      ceph_assert(unlink_ret == 0);
+      file_migrated_count++;
+    }
+    return -EIO;
+  }
+  
+  // everything went okay, rename files
+  file_migrated_count = 0;
+  for (auto &[name, file] : files_to_migrate) {
+    
+    dout(5) << fmt::format("{} unlinking {}", __func__, name) << dendl;
+    int unlink_ret = unlink(wal_dir, name);
+    ceph_assert(unlink_ret == 0);
+    
+    string new_file_temp_name = fmt::format("{}{}", tmp_name_prefix, file_migrated_count);
+    dout(5) << fmt::format("{} renaming {} to {}", __func__, tmp_name_prefix, name) << dendl;
+    int rename_ret = rename(wal_dir, new_file_temp_name, wal_dir, name);
+    ceph_assert(rename_ret == 0);
+    file_migrated_count++;
+  }
+  
+  // Ensure no dangling wal v2 files are inside transactions.
+  _compact_log_sync_LNF_LD();
+  
+  _write_super(BDEV_DB, 2);
```

Comment:
This is downgrade to 1, version shouldn't be 2 then?

Reply:
not by brightest day